Backfill guide (#203)
googleforgames/open-match-docs: site/content/en/docs/Guides/Backfill/_index.md

* A social or world place in a game allows players to see others in the space, but other players are not strictly required. When players enter the game it is preferred to put players into an existing server, preventing players from joining low population servers. As existing servers reach capacity, the game server management service starts up new empty servers. The new empty servers request players to be backfilled into them, allowing more players to join world servers. Servers may be very long living, as new players replace the leaving players.
* A multiplayer game has a player(s), leave the game, and makes a request for new players to be backfilled into the game. One of the players in the game invites a friend to join using a social platform. The friend joins the game, and the game server cancels its backfill request.

## Main algorithm

1. Player creates Ticket (__T1__).
2. Director calls Backend which calls MatchMaking Function (MMF) to build matches with the MatchProfile passed along. 

> 3. MMF finds __T1__, creates a new Backfill (__B1__*), assings __T1__ to __B1__ and sets __T1__ status as __Pending__. In the response to Director it is sent a Match with ticket __T1__ and backfill __B1__.

Reviewer: assigns __T1__

Author: fixed

* A social or world place in a game allows players to see others in the space, but other players are not strictly required. When players enter the game it is preferred to put players into an existing server, preventing players from joining low population servers. As existing servers reach capacity, the game server management service starts up new empty servers. The new empty servers request players to be backfilled into them, allowing more players to join world servers. Servers may be very long living, as new players replace the leaving players.
* A multiplayer game has a player(s), leave the game, and makes a request for new players to be backfilled into the game. One of the players in the game invites a friend to join using a social platform. The friend joins the game, and the game server cancels its backfill request.

## Main algorithm

1. Player creates Ticket (`T1`).
2. Director calls `Backend.FetchMatches` with a MatchProfile.
3. MMF runs `QueryBackfills` and `QueryTickets` using the MatchProfile. It returns `T1`.
4. MMF use `T1`, constructs a new Backfill (`B1`*) and returns a `Match` containing ticket `T1` and Backfill `B1`.

> 5. Director starts allocating GameServer.

Reviewer 2: ```suggestion
5. Director starts allocating GameServer with Backfill ID information.
```

> 5. Director starts allocating GameServer with Backfill `B1` information.

Reviewer 2: ```suggestion
5. Director starts allocating GameServer with Backfill `B1.Id` information.
```

Author: should be only focused on ID?? if eventually something is added to Backfill, the ID is misleading, in my opinion

6. Another player creates a Ticket (`T2`).
7. Director calls `Backend.FetchMatches`.
8. MMF runs `QueryBackfills` and `QueryTickets` using the MatchProfile. They return `B1` and `T2` accordingly.
9. MMF determines `B1` could be used because it has open slots available. `T2` and `B1` forms a new Match.

> 10. When GameServer allocating ends, it starts polling Open Match to acknowledge the backfill by ID received in the Match (`B1.ID`) and in order to assign the address of the GameServer for `T1` and `T2`.

Reviewer 2: Or this can be somewhere in the next PR.

> 10. When GameServer allocating ends, it starts polling Open Match to acknowledge the backfill by ID received in the Match (`B1.ID`) and in order to assign the address of the GameServer for `T1` and `T2`.

Author: i applied the changes in this PR

> 10. When GameServer allocating ends, it starts polling Open Match to acknowledge the backfill with `Backend.AcknowledgeBackfill` by ID received in the Match (`B1.ID`) and in order to assign the address of the GameServer for `T1` and `T2`.

Reviewer 2: Sorry it is Frontend
```suggestion
10. When GameServer allocating ends, it starts polling Open Match to acknowledge the backfill with `Frontend.AcknowledgeBackfill` by ID received in the Match (`B1.ID`) and in order to assign the address of the GameServer for `T1` and `T2`.
```

*: The Backfill should be created with field `AllocateGameServer` set to `true`, so GameServer orchestrator (e.g.: [Agones](https://agones.dev/site/)) knows it has to create a new GameServer.

## Considerations

Since Backfill is a large feature, using it requires changes to many components inside Open Match.
If Backfill is not being used, no changes are required since the Backfill API doesn't change 
the behavior of the current Open Match API state.

### Acknowledging Backfills

> To Acknowledge a Backfill is a request, that acts in form of a ping in an interval, made to Open Match Frontend to notify the associated tickets with their assignment.

Reviewer 2: ```suggestion
AcknowledgeBackfill is a request, that acts in form of a periodic ping, made to Open Match Frontend to set the assignment on the tickets associated with this backfill.
```

GameServers would run AcknowledgeBackfill every N seconds, where N is, at most, 80% of the duration set for `pendingReleaseTimeout` (N < 0.8 * pendingReleaseTimeout).

### Cleaning up Backfills

> This process, as its name suggests, deletes the Backfill objects from the state store that have a last acknowledge time bigger than the configured timeout duration.

Reviewer 2: ```suggestion
This process, as its name suggests, deletes expired Backfill objects from the state store, those Backfills not acknowledged for the aforementioned `backfillTTL` period.
```

Author: `This process deletes expired Backfill objects from the state store those which were not acknowledged for the aforementioned `backfillTTL` period.` what do you think @aLekSer ??

Reviewer 2: It is also fine for me.

Reviewer 2: This process deletes expired Backfill objects from the state store. Expired means those backfills which were not acknowledged for the aforementioned "backfillTTL" period.

> Obs: you need to keep present if time for acknowledging a backfill expires, you’ll need to recreate the Backfill object again.

Reviewer 2: ```suggestion
Note: you need to keep acknowledging a backfill and if it expires you need to recreate the Backfill object again.
```

8. MMF runs `QueryBackfills` and `QueryTickets` using the MatchProfile. They return `B1` and `T2` accordingly.
9. MMF determines `B1` could be used because it has open slots available. `T2` and `B1` forms a new Match.
10. When GameServer allocating ends, it starts polling Open Match to acknowledge the backfill with `Frontend.AcknowledgeBackfill` by ID received in the Match (`B1.ID`) and in order to assign the address of the GameServer for `T1` and `T2`.

*: The Backfill should be created with field `AllocateGameServer` set to `true`, so GameServer orchestrator (e.g.: [Agones](https://agones.dev/site/)) knows it has to create a new GameServer.

## Considerations

Since Backfill is a large feature, using it requires changes to many components inside Open Match.
If Backfill is not being used, no changes are required since the Backfill API doesn't change 
the behavior of the current Open Match API state.

### Acknowledging Backfills

AcknowledgeBackfill is a request, that acts as a periodic ping, made to Open Match Frontend to set the assignment on the tickets associated with this backfill.
GameServer receives current Backfill status as a response.

> GameServers would run AcknowledgeBackfill every N seconds, where N is, at most, 80% of the duration set for `pendingReleaseTimeout` (N < 0.8 * pendingReleaseTimeout).

Reviewer: "...80% of the duration set for `pendingReleaseTimeout` (N < 0.8 * pendingReleaseTimeout)." ->
80% of the duration set for `pendingReleaseTimeout`. backfillTTL = 0.8 * pendingReleaseTimeout. N < backfillTTL.

Just to introduce backfillTTL. What do you think?

Author: i would use `backfillTTL represents 80% of pendingReleaseTimeout and this value MUST always be bigger than N (backfillTTL = 0.8 * pendingReleaseTimeout; N < backfillTTL, where N is AcknowledgeBackfill interval)`. How is that??

Reviewer: Ok, let's leave it as you suggest/

### Cleaning up Backfills

> This process deletes expired Backfill objects from the state store. Expired means those backfills which were not acknowledged for the aforementioned `backfillTTL` period.

Reviewer: I do not see where backfillTTL is mentioned. 
I suggest the following rephrasing:
Instead of 
"Expired means those backfills which were not acknowledged for the aforementioned `backfillTTL` period." ->
"Backifll becomes expired if it is not acknowledged in "backfillTTL" period."

Author: applied 👍

1. Player creates Ticket (`T1`).
2. Director calls `Backend.FetchMatches` with a MatchProfile.
3. MMF runs `QueryBackfills` and `QueryTickets` using the MatchProfile. It returns `T1`.
4. MMF use `T1`, constructs a new Backfill (`B1`*) and returns a `Match` containing ticket `T1` and Backfill `B1`.
5. Director starts allocating GameServer with Backfill `B1` information.
6. Another player creates a Ticket (`T2`).
7. Director calls `Backend.FetchMatches`.
8. MMF runs `QueryBackfills` and `QueryTickets` using the MatchProfile. They return `B1` and `T2` accordingly.
9. MMF determines `B1` could be used because it has open slots available. `T2` and `B1` forms a new Match.
10. When GameServer allocating ends, it starts polling Open Match to acknowledge the backfill with `Frontend.AcknowledgeBackfill` by ID received in the Match (`B1.ID`) and in order to assign the address of the GameServer for `T1` and `T2`.

*: The Backfill should be created with field `AllocateGameServer` set to `true`, so GameServer orchestrator (e.g.: [Agones](https://agones.dev/site/)) knows it has to create a new GameServer.

## Considerations

> Since Backfill is a large feature, using it requires changes to many components inside Open Match.

Author: I took this paragraph from our design proposal document, but i can remove it if you think it is not appropriate here

> Since Backfill is a large feature, using it requires changes to many components inside Open Match.

Reviewer 2: @hsorellana please remove this sentence or rephraze.

---

## Background

Some scenarios that could happen when playing multiplayer games:

* A multiplayer game has a player(s) leave the game. To keep the game interesting and fun, matchmaking will replace the missing player with someone new.

> * A large player count game (e.g.: battleroyal) will create a non-full match and allow players to connect to the game server prior to it being full. Players will wait in a lobby on the game server while new players are backfilled into the game until the server is fully populated.

Reviewer 3: battleroyal -> battle royale

Author: fixed 👍
